crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,742 @@
+            },
+            None => CommutationLibrary {
+                library: Some(HashMap::new()),
+            },
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum CommutationLibraryEntry {
+    Commutes(bool),
+    QubitMapping(HashMap<SmallVec<[Option<Qubit>; 2]>, bool>),
+}
+
+impl<'py> FromPyObject<'py> for CommutationLibraryEntry {
+    fn extract_bound(b: &Bound<'py, PyAny>) -> Result<Self, PyErr> {
+        if let Ok(b) = b.extract::<bool>() {
+            return Ok(CommutationLibraryEntry::Commutes(b));
+        }
+        let dict = b.downcast::<PyDict>()?;
+        let mut ret = hashbrown::HashMap::with_capacity(dict.len());
+        for (k, v) in dict {
+            let raw_key: SmallVec<[Option<u32>; 2]> = k.extract()?;
+            let v: bool = v.extract()?;
+            let key = raw_key.into_iter().map(|key| key.map(Qubit)).collect();
+            ret.insert(key, v);
+        }
+        Ok(CommutationLibraryEntry::QubitMapping(ret))
+    }
+}
+
+impl ToPyObject for CommutationLibraryEntry {
+    fn to_object(&self, py: Python) -> PyObject {
+        match self {
+            CommutationLibraryEntry::Commutes(b) => b.into_py(py),
+            CommutationLibraryEntry::QubitMapping(qm) => {
+                let out_dict = PyDict::new_bound(py);
+
+                qm.iter().for_each(|(k, v)| {
+                    out_dict
+                        .set_item(
+                            PyTuple::new_bound(
+                                py,
+                                k.iter()
+                                    .map(|q| q.map(|t| t.0))
+                                    .collect::<Vec<Option<u32>>>(),
+                            ),
+                            PyBool::new_bound(py, *v),
+                        )
+                        .ok()
+                        .unwrap()
+                });
+                out_dict.unbind().into_any()
+            }
+        }
+    }
+}
+
+type CacheKey = (
+    SmallVec<[Option<Qubit>; 2]>,
+    (SmallVec<[ParameterKey; 3]>, SmallVec<[ParameterKey; 3]>),
+);
+// Need a struct instead of a type definition because we cannot implement serialization traits otherwise
+#[derive(Clone, Debug)]
+struct CommutationCacheEntry {
+    mapping: HashMap<CacheKey, bool>,
+}
+impl CommutationCacheEntry {
+    fn get(&self, key: &CacheKey) -> Option<&bool> {
+        self.mapping.get(key)
+    }
+    fn iter(&self) -> Iter<'_, CacheKey, bool> {
+        self.mapping.iter()
+    }
+
+    fn insert(&mut self, k: CacheKey, v: bool) -> Option<bool> {
+        self.mapping.insert(k, v)
+    }
+}
+
+impl ToPyObject for CommutationCacheEntry {
+    fn to_object(&self, py: Python) -> PyObject {
+        let out_dict = PyDict::new_bound(py);
+        for (k, v) in self.iter() {
+            let qubits = PyTuple::new_bound(
+                py,
+                k.0.iter()
+                    .map(|q| q.map(|t| t.0))
+                    .collect::<Vec<Option<u32>>>(),
```

Comment:
Avoid collecting here, let the `PyTuple` do the collection.
```suggestion
            let qubits = PyTuple::new_bound(
                py,
                k.0.iter()
                    .map(|q| q.map(|t| t.0))
```